temporal-sdk/src/main/java/io/temporal/internal/replay/ServiceWorkflowHistoryIterator.java

```diff
@@ -0,0 +1,142 @@
+      PollWorkflowTaskQueueResponseOrBuilder task,
+      Scope metricsScope) {
+    this.service = service;
+    this.namespace = namespace;
+    this.task = task;
+    this.metricsScope = metricsScope;
+    // TODO Refactor WorkflowHistoryIteratorTest or WorkflowHistoryIterator to remove this check.
+    //  `service == null` shouldn't be allowed as it's needed for a normal functioning of this
+    // class.
+    this.grpcRetryer = service != null ? new GrpcRetryer(service.getServerCapabilities()) : null;
+    History history = task.getHistory();
+    current = history.getEventsList().iterator();
+    nextPageToken = task.getNextPageToken();
+  }
+
+  // Returns true if more history events are available.
+  // Server can return page tokens that point to empty pages.
+  // We need to verify that page is valid before returning true.
+  // Otherwise next() method would throw NoSuchElementException after hasNext() returning true.
+  @Override
+  public boolean hasNext() {
+    if (current.hasNext()) {
+      return true;
+    }
+    if (nextPageToken.isEmpty()) {
+      return false;
+    }
+
+    GetWorkflowExecutionHistoryResponse response = queryWorkflowExecutionHistory();
+
+    current = response.getHistory().getEventsList().iterator();
+    nextPageToken = response.getNextPageToken();
+
+    return current.hasNext();
+  }
+
+  @Override
+  public HistoryEvent next() {
+    if (hasNext()) {
+      return current.next();
+    }
+    throw new NoSuchElementException();
+  }
+
+  public void initDeadline(Deadline deadline) {
+    this.deadline = deadline;
+  }
+
+  GetWorkflowExecutionHistoryResponse queryWorkflowExecutionHistory() {
+    RpcRetryOptions retryOptions =
+        RpcRetryOptions.newBuilder()
+            .setInitialInterval(retryServiceOperationInitialInterval)
+            .setMaximumInterval(retryServiceOperationMaxInterval)
+            .build();
+    GrpcRetryer.GrpcRetryerOptions grpcRetryerOptions =
+        new GrpcRetryer.GrpcRetryerOptions(retryOptions, deadline);
+    GetWorkflowExecutionHistoryRequest request =
+        GetWorkflowExecutionHistoryRequest.newBuilder()
+            .setNamespace(namespace)
+            .setExecution(task.getWorkflowExecution())
+            .setNextPageToken(nextPageToken)
+            .build();
+    try {
+      return grpcRetryer.retryWithResult(
+          () ->
+              service
+                  .blockingStub()
+                  .withOption(METRICS_TAGS_CALL_OPTIONS_KEY, metricsScope)
+                  .getWorkflowExecutionHistory(request),
+          grpcRetryerOptions);
+    } catch (StatusRuntimeException ex) {
+      if (Status.DEADLINE_EXCEEDED.equals(ex.getStatus())) {
+        throw Status.DEADLINE_EXCEEDED
+            .withDescription(
+                "getWorkflowExecutionHistory pagination took longer than workflow task timeout")
+            .asRuntimeException();
+      }
+      throw ex;
+    } catch (Exception e) {
+      // TODO Error? Why?
```

Comment:
Needs cleanup here I think

Reply:
This is an old code, I have a separate task to clean this up